a10networks/a10-octavia
# STACK-1375, STACK-1376, STACK-1377 Changes from PP sessions with team

## Description
a) In the VRRP-A High Availability & aVCS environment if one of the device in the cluster goes down then the configuration changes are handled gracefully without erring out. When the other device of the cluster comes up, the configurations to the device that came up can be pushed through update calls.
b) DCHP -> Static & Static -> DHCP config changes fail. Neutron ports are created, updated and deleted according to config.

## Jira Ticket
https://a10networks.atlassian.net/browse/STACK-1375
https://a10networks.atlassian.net/browse/STACK-1376
https://a10networks.atlassian.net/browse/STACK-1377
[STACK-1347](https://a10networks.atlassian.net/browse/STACK-1347)
[STACK-1349](https://a10networks.atlassian.net/browse/STACK-1349)



## Technical Approach
a)A new task SetupDeviceNetworkMap is introduced to setup device_network_map in vthunder to contain only vcs devices with known states and have Master device object at index 0. The vcs-summary information is obtained from the Master and used to determine the device status in the cluster. Later the TagInterface* tasks act upon only up devices.

b) Updation Approach
- Since ACOS client api `interface.ve.update()` doesn't seem to work in case of static ve_ip to dhcp updation. It throws error 
`DHCP address has to be the first configured under an interface.`

Hence used approach of first `interface.ve.delete()` then `interface.ve.create()`.

## Config Changes
<pre>
[a10_global]
network_type = "vlan"

[hardware_thunder]
devices=[
 {
 "project_id": "9b990719b64a4b2897d378ba56bba78d",
 "ip_address": "10.0.0.99",
 "device_name": "rack_vthunder",
 "username": "admin",
 "password": "a10",
 "interface_vlan_map": {
  "device_1": {
   "vcs_device_id": 1,
   <b>"mgmt_ip_address": "10.0.0.74",</b>
   "ethernet_interfaces": [{
    "interface_num": 2,
    "vlan_map": [
     {"vlan_id": 12, "ve_ip": ".10"}
    ]
   }],
   "trunk_interfaces": [{
    "interface_num": 1,
    "vlan_map": [
     {"vlan_id": 13, "ve_ip" : ".10"},
     {"vlan_id": 14, "ve_ip": ".10"}
    ]
   }],
  },
  "device_2": {
   "vcs_device_id": 2,
   <b>"mgmt_ip_address": "10.0.0.47",</b>
   "ethernet_interfaces": [{
    "interface_num": 2,
    "vlan_map": [
     {"vlan_id": 12, "ve_ip": ".20"}
    ]
   }],
   "trunk_interfaces": [{
    "interface_num": 1,
    "vlan_map": [
     {"vlan_id": 13, "ve_ip" : ".20"},
     {"vlan_id": 14, "ve_ip": ".20"}
    ]
   }],
  }
 }
 }]
</pre>

## Test Cases
a) Unit test cases changes contain positive and negative test cases.
b) **_For ve_ip creates, updates and delete_**
CREATION on vThunder and openstack neutron ports
- if a static `ve_ip` is configured in `ethernet_interfaces` -> CREATED
- if `use_dhcp` set as `True` in `ethernet_interfaces` -> CREATED
- if static `ve_ip` is configured in `trunk_interfaces` -> CREATED
- if `use_dhcp` set as `True` in `trunk_interfaces` -> CREATED


UPDATION on vThunder and openstack neutron ports
- if existing static `ve_ip` is changed to new `ve_ip` in `ethernet_interfaces` -> UPDATED
- if existing static `ve_ip` is changed to dhcp setting in `ethernet_interfaces` -> UPDATED
- if existing dhcp is changed to static ve_ip in `ethernet_interfaces` -> UPDATED
- if existing static `ve_ip` is changed to new `ve_ip` in `trunk_interfaces` -> UPDATED
- if existing static `ve_ip` is changed to dhcp setting in `trunk_interfaces` -> UPDATED
- if existing dhcp is changed to static ve_ip in `trunk_interfaces` -> UPDATED

DELETION on vThunder and openstack neutron ports
- if member is deleted, corresponding vlan, ve/trunk ve_ip configs are deleted 

## Manual Testing

a) device 1 and device 2 testing
1) State device-1: Master and device-2: vBlade
   1a) Add/Update/Delete Loadbalancer, check applies to both devices.
       $ openstack loadbalancer create --provider=a10 --name a10lb1 --vip-subnet-id provider-vlan-12-subnet --project 9b990719b64a4b2897d378ba56bba78d
       # Update VLAN 12 device-1 ve_ip: ".11" and device_2 ve_ip: ".21"
       # sudo systemctl restart a10-controller-worker.service
       $ openstack loadbalancer set a10lb1
       $ openstack loadbalancer delete a10lb1
   1b) Add/Update/Delete Member, check applies to both devices
       $ openstack loadbalancer create --provider=a10 --name a10lb1 --vip-subnet-id provider-vlan-12-subnet --project 9b990719b64a4b2897d378ba56bba78d
       $ openstack loadbalancer listener create --name listener1 --protocol HTTP --protocol-port 80 a10lb1
       $ openstack loadbalancer pool create --name pool1 --lb-algorithm ROUND_ROBIN --listener listener1 --protocol HTTP
       $ openstack loadbalancer member create --subnet-id provider-vlan-13-subnet --address  10.0.13.110 --protocol-port 80 pool1
       # Update VLAN 13 device-1 ve_ip: ".11" and device_2 ve_ip: ".21"
       # sudo systemctl restart a10-controller-worker.service
       $ openstack loadbalancer member set pool1 f30d983d-e488-4afe-ad45-8b6f49b2d217
       $ openstack loadbalancer member delete pool1 f30d983d-e488-4afe-ad45-8b6f49b2d217
       $ openstack loadbalancer pool delete pool1
       $ openstack loadbalancer listener delete listener1
       $ openstack loadbalancer delete a10lb1

2) Make device-2 down(link down). State device-1: Master and device-2: Unknown
   $ sudo ip link set vnet5 down
   2a) Add/Update/Delete Loadbalancer, check applies to one device.
       #commands same as "1a"
   2b) Add/Update Member, check applies to one device.
       #commands same as "1b" without deletes
       $ openstack loadbalancer member set pool1 ad2094db-8ccb-40bb-ab7f-1c45fe20352a

3) Bring device-2 up State device-1: Master and device-2: vBlade
   $ sudo ip link set vnet5 up
   3a) Update LB and Member, check device-1 config is synced to device-2
       $ openstack loadbalancer set a10lb1
       $ openstack loadbalancer member set pool1 ad2094db-8ccb-40bb-ab7f-1c45fe20352a

4) Make device-1 down(Reload) State device-1: Unknown and device-2: Master
   4a) Add one more Member on device-2
       $ openstack loadbalancer member create --subnet-id provider-vlan-14-subnet --address  10.0.14.110 --protocol-port 80 pool1

5) Bring device-1 up State device-1: vBlade and device-2: Master
   5a) Update Member, check device-2 config is synced to device-1
       $ openstack loadbalancer member set pool1 c29147a2-21ac-4348-856a-6c1b5bd9964f
   5b) Add one more member and check config
       $ openstack loadbalancer member create --subnet-id provider-vlan-14-subnet --address  10.0.14.111 --protocol-port 80 pool1
   5c) Delete all config
       $ openstack loadbalancer member delete pool1 01d1e537-b399-42a4-97a0-4e982526b493
       $ openstack loadbalancer member delete pool1 c29147a2-21ac-4348-856a-6c1b5bd9964f
       $ openstack loadbalancer member delete pool1 ad2094db-8ccb-40bb-ab7f-1c45fe20352a
       $ openstack loadbalancer pool delete pool1
       $ openstack loadbalancer listener delete listener1
       $ openstack loadbalancer delete a10lb1

b) Prerequisite : 
- Use `install-a10-octavia` to install vlan setup.
- Have required ovs-vsctl setting. eg creating bonds etc.
- On aVCS setup, in vthunders, have set up vrid floating ip
- In `a10-octavia.conf`, include `vrid_floating_ip` and `standby_ip_address` setting.

STEP 1: Configure required ve_ip/dhcp setting for `ethernet_interfaces` or/and `trunk_interfaces` in `a10-octavia.conf`
STEP 2: Restart `a10-controller-worker.service` service
STEP 3: Create loadbalancer, listener, pool and member. 
STEP 4: Check on vThunder, if proper ve/trunk ips are configured.
STEP 5: Check on openstack UI for neutron port created with same ve_ip addresses.
